opt: zigzag scan hints

cucaroach opened the pull request:
Fixes #35814

Enable table@{FORCE_ZIGZAG} and table@{FORCE_ZIGZAG=index} hints to lower the cost
of zigzag plans to that they will be preferred over other plans.

Note that this is a breaking change, if customer has an index called
FORCE_ZIGZAG and is hinting with table@{FORCE_ZIGZAG} that will no longer do what
it did before. Not sure what the ramifications of that are.

Release note (sql change): Extend index scan hint to allow zigzag joins to be preferred.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67737)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Nice! The main feedback is that we actually want this "hint" to be a requirement. I guess the name "hint" is a bit misleading, but all the other index and join hints we support work this way: If a particular plan is hinted, we must use that plan or return an error if it's not possible.

> Note that this is a breaking change, if customer has an index called zigzag ...

One option could be to change the hint to `FORCE_ZIGZAG`, which seems less likely to be used as a name

> Reviewed 7 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-Mf3N97pEeZUBGludmo2:b-8rxdih)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @RaduBerinde)

___
*[pkg/sql/opt/xform/coster.go, line 1063 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mf3NxM34pyFmkijtcYv:-Mf3NxM34pyFmkijtcYw:b7xkz7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bcf8604bcfdd022ca1718e1c2e4f74ebc600dd4/pkg/sql/opt/xform/coster.go#L1063)):*
> ```Go
> 	cost += filterSetup
> 
> 	if join.HintProvided {
> ```

This is a different kind of hint, and not what we'll want to use here. I think you'll actually want to update this section, to return `hugeCost` if we plan a scan when a zigzag join was hinted: https://github.com/cockroachdb/cockroach/blob/148a9ec986042bde73b3f391c9632655ecfcbaba/pkg/sql/opt/xform/coster.go#L606-L610.

Then you'll also want to return `hugeCost` from `computeZigzagJoinCost` if the hinted indexes are not used.

You'll probably also need to update the code in `execbuilder` to return an error if the best plan used a scan when a zigzag join was forced.

Basically, if a user uses the ZIGZAG join flag, we want to force that plan, and return an error if it's not possible to plan the zig zag join.
___
*[pkg/sql/sem/tree/select.go, line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mf3QME5-BSJnkEFEoe1:-Mf3QME5-BSJnkEFEoe2:bpk92kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bcf8604bcfdd022ca1718e1c2e4f74ebc600dd4/pkg/sql/sem/tree/select.go#L291)):*
> ```Go
> 	IgnoreUniqueWithoutIndexKeys bool
> 	// HintZigzag enables costing hints to prefer a zigzag join.
> 	HintZigzag bool
> ```

I think we want this to be `ForceZigzag`


<!-- Sent from Reviewable.io -->

cucaroach commented:
RFAL!   I went a little sideways to try to prevent increasing memo sizes all over the place, let me know if its too weird.   I think I got the semantics @rytaft was after working but not sure if there's any other ways a scan can be planned that I need to error from if zigzag was forced.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEa9ip94_6mEdmTfqw:-MfEa9iqD8vykL5ZBdhp:bj5234u) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/memo/expr.go#L358)):*
> ```Go
> 	Direction  tree.Direction
> 
> 	// index is a uint32 to allow optimal packing of this structure, use Index()
> ```

You can consider making `flags` a pointer in `ScanPrivate` instead. You can have a singleton instance with zero flags that is used when there are no flags (the common case). The same flags structure can be shared between multiple Scans (eg in 
___
*[pkg/sql/opt/memo/expr.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbwIzBaUqljm7lZg6:-MfEbwIzBaUqljm7lZg7:b2g2xc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/memo/expr.go#L370)):*
> ```Go
> 	// For cases 2 and 3 if we can't generate a suitable plan an error will
> 	// result.
> 	ZigzagIndices *util.FastIntSet
> ```

Consider having a ForceZigzag flag here as well instead of the pointer (if it's next to the other bools, it shouldn't increase the struct size).
___
*[pkg/sql/sem/tree/select.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbKuy-aIZbcZRm-jp:-MfEbKuy-aIZbcZRm-jq:bnw1x73) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/sem/tree/select.go#L293)):*
> ```Go
> 	// There are three possibilities:
> 	// 1) nil means no hint, business as usual
> 	// 2) empty slice means force a Zigzag join but any index will do
> ```

Differentiating between nil and empty slice is pretty error prone. I'd consider adding a `ForceZigZag bool` flag.


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEa9ip94_6mEdmTfqw:-MfEfUgm4NkJ3t-ael_L:bfpbfwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/memo/expr.go#L358)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

You can consider making `flags` a pointer in `ScanPrivate` instead. You can have a singleton instance with zero flags that is used when there are no flags (the common case). The same flags structure can be shared between multiple Scans (eg in 
</blockquote></details>

Great idea, that will affect the memo memory usage sizes in a good way no doubt.
___
*[pkg/sql/opt/memo/expr.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbwIzBaUqljm7lZg6:-MfEfdE22e5vQA2NiG1A:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/memo/expr.go#L370)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Consider having a ForceZigzag flag here as well instead of the pointer (if it's next to the other bools, it shouldn't increase the struct size).
</blockquote></details>

:+1: 
___
*[pkg/sql/opt/xform/coster.go, line 1063 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mf3NxM34pyFmkijtcYv:-MfEflPLAYy15L9e6aHl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bcf8604bcfdd022ca1718e1c2e4f74ebc600dd4/pkg/sql/opt/xform/coster.go#L1063)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This is a different kind of hint, and not what we'll want to use here. I think you'll actually want to update this section, to return `hugeCost` if we plan a scan when a zigzag join was hinted: https://github.com/cockroachdb/cockroach/blob/148a9ec986042bde73b3f391c9632655ecfcbaba/pkg/sql/opt/xform/coster.go#L606-L610.

Then you'll also want to return `hugeCost` from `computeZigzagJoinCost` if the hinted indexes are not used.

You'll probably also need to update the code in `execbuilder` to return an error if the best plan used a scan when a zigzag join was forced.

Basically, if a user uses the ZIGZAG join flag, we want to force that plan, and return an error if it's not possible to plan the zig zag join.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 291 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mf3QME5-BSJnkEFEoe1:-MfEfpKG917EbTkiyo5D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bcf8604bcfdd022ca1718e1c2e4f74ebc600dd4/pkg/sql/sem/tree/select.go#L291)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think we want this to be `ForceZigzag`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbKuy-aIZbcZRm-jp:-MfEfqOMAEDekqyLF1cF:b-d5b9w3) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/sem/tree/select.go#L293)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Differentiating between nil and empty slice is pretty error prone. I'd consider adding a `ForceZigZag bool` flag.
</blockquote></details>

:+1: It felt dirty


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once you address the feedback from @raduberinde 

> Reviewed 2 of 19 files at r1, 21 of 21 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MfYCBmT-cmHal3LZdQj:b33933c)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @RaduBerinde)

___
*[pkg/sql/opt/xform/select_funcs.go, line 1087 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfYAwbS4yG3YHn6upYY:-MfYAwbT-6kfNXUDfaki:b5vljtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/xform/select_funcs.go#L1087)):*
> ```Go
> 				indices := util.MakeFastIntSet(leftIndex.Ordinal(), rightIndex.Ordinal())
> 				forceIndices := scanPrivate.Flags.ZigzagIndices
> 				if !forceIndices.Intersection(indices).Equals(*forceIndices) {
> ```

Isn't this equivalent to `if !forceIndices.SubsetOf(indices)`?


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
RFAL, I followed Radu's suggestion but it didn't lead me where I thought it would.    Go doesn't have proper constructors and apparently ScanPrivate's get ginned up in the bowels of reflection in exprgen and in test code so I couldn't figure out how to make the install default pointer approach work.   So I just made the nil checking approach more palatable with helpers.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MfbHz2M9ax-ZHKPrTEm:b-fj03n8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/xform/select_funcs.go, line 1087 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfYAwbS4yG3YHn6upYY:-MfbIfiO9eSIgWWcqcde:bmiskcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/opt/xform/select_funcs.go#L1087)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Isn't this equivalent to `if !forceIndices.SubsetOf(indices)`?
</blockquote></details>

Yes, good idea.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 6 of 16 files at r3, 16 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/zigzag_join, line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mfc-SJHAGdXa2BbA-Dn:-Mfc-SJHAGdXa2BbA-Do:b-mlxuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/logictest/testdata/logic_test/zigzag_join#L220)):*
> ```
> SELECT * FROM a@{FORCE_ZIGZAG=a_idx,FORCE_ZIGZAG=a_idx} WHERE a = 3 AND b = 7;
> 
> # Full indexs specified.
> ```

nit: indexs -> indexes
___
*[pkg/sql/opt/memo/expr.go, line 360 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mfc-mKuAsJWeEhNR9W_:-Mfc-mKuAsJWeEhNR9Wa:bs0jrru) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L360)):*
> ```Go
> 	Index       int
> 
> 	// ZigzagIndices makes planner to prefer a zigzag with particular indices.
> ```

nit: makes planner to prefer -> makes planner prefer
___
*[pkg/sql/opt/memo/expr.go, line 691 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mfc23g6964_qvKflJ6P:-Mfc23g6964_qvKflJ6Q:bsayfzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L691)):*
> ```Go
> // Some ScanFlags helpers to deal with nil checking.  Attempts were made to
> // implant a pointer to a global "default" ScanFlags but because of reflection
> // usage I couldn't figure out how to make that work seamlessly.
> ```

nit: it would be helpful to have a `TODO(treilly): ...` here so future readers can ask you about it if needed.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
I think you should also update `memo/expr_format.go` to print the zigzag hint if it's provided (and make sure it shows up in some test output).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MfcE2WFDnrhWLlMyLek:bqxo9m4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/expr_format.go, line 413 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcELIc08OUkaHbf2-D:-MfcELIc08OUkaHbf2-E:b-xs9kp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b018ef444c4c740a27e7050dffef34ed0dab759b/pkg/sql/opt/memo/expr_format.go#L413)):*
> ```Go
> 				}
> 				tp.Childf("flags: force-index=%s%s", idx.Name(), dir)
> 			}
> ```

If the zigzag hint is set, you should print it here


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> RFAL, I followed Radu's suggestion but it didn't lead me where I thought it would. Go doesn't have proper constructors and apparently ScanPrivate's get ginned up in the bowels of reflection in exprgen and in test code so I couldn't figure out how to make the install default pointer approach work. So I just made the nil checking approach more palatable with helpers.

Can we keep `ScanPrivate.Flags` as a non-pointer then? I'm not convinced the memory savings in memos is worth the complexity added for nil-checking, interning, etc. If we are worried about the memo size increasing, then we could represent the 3 bools in the struct as bits in a single `int8` and add some helpers for getting the values and setters that uphold the invariants like "ForceIndex and NoIndexJoin cannot be set at the same time".

> Reviewed 3 of 19 files at r1, 1 of 21 files at r2, 6 of 16 files at r3, 16 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MfcFqYU1l_45m3muShH:b-9s2zvm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 354 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcD1ua2GrtxpAj415S:-MfcD1ua2GrtxpAj415T:b-afic2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L354)):*
> ```Go
> 
> 	// ForceIndex forces the use of a specific index (specified in Index).
> 	// ForceIndex and NoIndexJoin cannot both be set at the same time.
> ```

I'm assuming ForceIndex and ForceZigzag also cannot be used at the same time? We should make that clear here and add a test for it.

What about NoIndexJoin with ForceZigzag? Maybe add some tests for that combination too.
___
*[pkg/sql/opt/memo/expr.go, line 691 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mfc23g6964_qvKflJ6P:-MfcFe1l3-6s2xvdCFvj:b-sa7vkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L691)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: it would be helpful to have a `TODO(treilly): ...` here so future readers can ask you about it if needed.
</blockquote></details>

I'm not convinced that making `ScanPrivate.Flags` a pointer is 
___
*[pkg/sql/sem/tree/select.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbKuy-aIZbcZRm-jp:-MfcHDHX0gyp9c_TdjDB:b-a278af) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/sem/tree/select.go#L293)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

:+1: It felt dirty
</blockquote></details>

Looks like this still needs to be addressed?


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/expr.go, line 354 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcD1ua2GrtxpAj415S:-MfcTKa8B-81yt_8q-I7:b-32lpn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L354)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm assuming ForceIndex and ForceZigzag also cannot be used at the same time? We should make that clear here and add a test for it.

What about NoIndexJoin with ForceZigzag? Maybe add some tests for that combination too.
</blockquote></details>

Yeah, I should error out if any of these are combined, nice catch.
___
*[pkg/sql/opt/memo/expr.go, line 691 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-Mfc23g6964_qvKflJ6P:-MfcRzeJCzTK80TxPuog:b-iabuxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L691)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm not convinced that making `ScanPrivate.Flags` a pointer is 
</blockquote></details>

ScanFlags was 16 bytes, by adding the FastIntSet I think its up to 32.   So the nil approach saves us 24 bytes per ScanExpr in the common case where ScanFlags is empty.   But this is not the hill I want to die on.   Happy to put it back to a plain struct if we don't think its worth it...
___
*[pkg/sql/sem/tree/select.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfEbKuy-aIZbcZRm-jp:-MfcJa1n96qSzkiEbFvv:b-7wnisb) ([raw file](https://github.com/cockroachdb/cockroach/blob/263eed9c62bfa106231853124f79bfd2aea273e3/pkg/sql/sem/tree/select.go#L293)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like this still needs to be addressed?
</blockquote></details>

Oh right, I fixed the ScanFlags but forgot about IndexFlags.     Will fix.
___
*[pkg/sql/opt/memo/expr_format.go, line 413 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcELIc08OUkaHbf2-D:-MfcRXUxA48AFg_NmMMw:b-i42zab) ([raw file](https://github.com/cockroachdb/cockroach/blob/b018ef444c4c740a27e7050dffef34ed0dab759b/pkg/sql/opt/memo/expr_format.go#L413)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

If the zigzag hint is set, you should print it here
</blockquote></details>

This code is only reached for ScanExpr so that will never happen right?    I could print if below for ZigzagExprs, I'd just have to dupe the ScanFlags from ScanPrivate into ZigzagJoinPrivate, should I do that?


<!-- Sent from Reviewable.io -->

rytaft commented:
Can you also add some tests for inverted zigzag joins? You might also need to update `GenerateInvertedZigzagJoins`.





<!-- Sent from Reviewable.io -->

rytaft commented:
(sorry -- `GenerateInvertedIndexZigzagJoins`)





<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/xform/testdata/rules/join, line 8643 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfeyNN10WbSZKTJarc_:-MfeyNN10WbSZKTJarca:b-vthoj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/xform/testdata/rules/join#L8643)):*
> ```
>       └── name:2 = 'cpu' [outer=(2), constraints=(/2: [/'cpu' - /'cpu']; tight), fd=()-->(2)]
> 
> exec-ddl
> ```

nit: these tests should go in xform/testdata/rules/select with the other zigzag join tests


<!-- Sent from Reviewable.io -->

rytaft commented:
Sorry for the trickle of comments... would be good to add a test in `sql/parser/testdata/select_clauses` too





<!-- Sent from Reviewable.io -->

cucaroach committed:
opt: extend hints to Zigzag joins

Fixes #35814

Enable table@{FORCE_ZIGZAG} and table@{FORCE_ZIGZAG=index} hints to
force zigzag plans to be used.   Generate an error if a zigzag can not
be planned.

Note that this is a breaking change, if customer has an index called
FORCE_ZIGZAG and is hinting with table@{FORCE_ZIGZAG} that will no
longer do what it did before.

Release note (sql change): Extend index scan hint to allow zigzag joins
to be forced.

cucaroach reviewed (commented):
RFAL, I kiboshed any attempts at memory savings, if we feel strongly about that we can address with a separate PR.   I also added some sql parser tests and dealt with inverted zigzag joins.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-Mg6FRaY0tjrZUqzc1iS:b44202n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/xform/testdata/rules/join, line 8643 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfeyNN10WbSZKTJarc_:-Mg6Ye507nCkD432XC0R:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/xform/testdata/rules/join#L8643)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: these tests should go in xform/testdata/rules/select with the other zigzag join tests
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 28 of 28 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/zigzag_join, line 203 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHHqrS6gO9OOuKXt2h:-MgHHqrS6gO9OOuKXt2i:b-jarsjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/logictest/testdata/logic_test/zigzag_join#L203)):*
> ```
> 
> statement error syntax error
> SELECT * FROM a@{FORCE_ZIGZAG=foo,bar} WHERE a = 3 AND b = 7;
> ```

nit: these syntax tests probably aren't necessary here if you have these cases in `pkg/sql/parser/testdata` tests.
___
*[pkg/sql/logictest/testdata/logic_test/zigzag_join, line 220 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHI8nQEAM2QCvSHAVM:-MgHI8nQEAM2QCvSHAVN:b-pkf7yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/logictest/testdata/logic_test/zigzag_join#L220)):*
> ```
> # Basic form w/o indices.
> statement ok
> SELECT * FROM a@{FORCE_ZIGZAG} WHERE a = 3 AND b = 7;
> ```

nit: add a test where a zig zag can't be planned like `SELECT * FROM a@{FORCE_ZIGZAG} WHERE a = 3 AND c = 'foo'`
___
*[pkg/sql/opt/memo/expr_format.go, line 430 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHMHwA2Ht447fMKITf:-MgHMHwA2Ht447fMKITg:b-x80be) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/expr_format.go#L430)):*
> ```Go
> 						b.WriteString(string(idx.Name()))
> 						if _, needComma := s.Next(i + 1); needComma {
> 							b.WriteByte(',')
> ```

nit: rather than looking ahead you could print the comma before the index name for each index except the first.

```
needComma := false
for ... {
   if needComma {
       b.WriteByte(',')
    }
    b.WriteString(...)
    needComma = true
}
```
___
*[pkg/sql/opt/memo/testdata/format, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHN3c25AnuXdCXSTWu:-MgHN3c25AnuXdCXSTWv:b-st68j) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/testdata/format#L352)):*
> ```
> # Verify naked hint.
> opt
> SELECT * FROM zigzag@{FORCE_ZIGZAG} WHERE a = 3 AND b = 7;
> ```

 There's no force_zigzag printed because they are only printed for scan expressions. To test your changes in `expr_format.go` you need scans to show up in the plan. You could either remove the secondary indexes from the `zigzag` table or use the `norm` directive instead of the `opt` directive so that there is no exploration performed (or both!).
___
*[pkg/sql/opt/optbuilder/select.go, line 540 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHJBog-S2Pf676Jg6u:-MgHJBog-S2Pf676Jg6v:bsxvzz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/optbuilder/select.go#L540)):*
> ```Go
> 		private.Flags.ForceZigzag = indexFlags.ForceZigzag
> 		if indexFlags.ZigzagIndices != nil {
> 			private.Flags.ForceZigzag = true
> ```

Would it be simpler if the parser set `indexFlags.ForceZigzag` to true in the `FORCE_ZIGZAG=foo` case? Then the fields in these flags map directly to the fields in the AST struct.
___
*[pkg/sql/opt/xform/select_funcs.go, line 1093 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHKOVX3zEs6x-6ZNj0:-MgHKOVX3zEs6x-6ZNj1:brfunml) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/select_funcs.go#L1093)):*
> ```Go
> 				forceIndices := scanPrivate.Flags.ZigzagIndices
> 				if !forceIndices.SubsetOf(indices) {
> 					return
> ```

Can we move this into 2 checks at the top of the `ForEach` and `ForEachStartingAfter` callbacks to avoid unnecessary work?
___
*[pkg/sql/opt/xform/select_funcs.go, line 1447 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHK2DtBtNX_GaTBSPA:-MgHK2DtBtNX_GaTBSPB:b-b4i6du) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/select_funcs.go#L1447)):*
> ```Go
> 
> 		if !scanPrivate.Flags.ZigzagIndices.Empty() && !scanPrivate.Flags.ZigzagIndices.Contains(index.Ordinal()) {
> 			return
> ```

Can we put this at the top of this `ForEach` callback to avoid unnecessary work in this case?
___
*[pkg/sql/opt/xform/testdata/coster/join, line 998 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHKf7nC9O97gnQXQLU:-MgHKf7nC9O97gnQXQLV:b-yiugac) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/testdata/coster/join#L998)):*
> ```
> ----
> 
> # Verify huge cost with bogus hint.
> ```

nit: make this a little more clear like: "Verify that a huge cost is added to a scan when there is a zigzag join hint"
___
*[pkg/sql/opt/xform/testdata/rules/join, line 8643 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfeyNN10WbSZKTJarc_:-MgHL3SvAEWjkHV6YsIK:b-sg6fym) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/xform/testdata/rules/join#L8643)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Looks like these tests still need to be moved to `pkg/sql/opt/xform/testdata/rules/select`. We have zigzag join tests there because the rules for generating zigzag joins operate on Select expressions, not on joins. 
___
*[pkg/sql/parser/sql.y, line 9577 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHF2Yq2H8W5AB7nK9v:-MgHF2Yq2H8W5AB7nK9w:b84aevr) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/parser/sql.y#L9577)):*
> ```Yacc
>   }
> |
>   FORCE_ZIGZAG '=' index_name
> ```

I think we should be consistent with `FORCE_INDEX` and handle a syntax like `FORCE_ZIGZAG '=' '[' iconst64 ']'` which allows forcing a specific index by its descriptor ID. There's some parser tests in `pkg/sql/parser/testdata/table_by_id` which you can add on to.
___
*[pkg/sql/parser/testdata/select_clauses, line 424 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHHSLXBWNLyM8jm9K5:-MgHHSLXBWNLyM8jm9K6:b-s7qgqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/parser/testdata/select_clauses#L424)):*
> ```
> at or near "}": syntax error: FORCE_ZIGZAG specified multiple times
> DETAIL: source SQL:
> SELECT a FROM foo@{FORCE_ZIGZAG,FORCE_ZIGZAG}
> ```

What is expected with `foo@{FORCE_ZIGZAG,FORCE_ZIGZAG=a}`? An error? Can you add a test for that case and maybe `foo@{FORCE_ZIGZAG=a,FORCE_ZIGZAG}` as well.
___
*[pkg/sql/sem/tree/select.go, line 293 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHGYkR4GF6H478qLl7:-MgHGYkR4GF6H478qLl8:bz7ctcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/sem/tree/select.go#L293)):*
> ```Go
> 	// index constraints.
> 	IgnoreUniqueWithoutIndexKeys bool
> 	ForceZigzag                  bool
> ```

nit: Add a comment explaining what this does and how it relates to `ZigzagIndices`.
___
*[pkg/sql/sem/tree/select.go, line 296 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHGgEL5ztw-51KyXje:-MgHGgEL5ztw-51KyXjf:b-ut0cbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/sem/tree/select.go#L296)):*
> ```Go
> 	// ZigzagIndices forces a zigzag plan, with (optionally) particular indices.
> 	// There are three possibilities:
> 	// 1) nil means no hint, business as usual
> ```

nit: update this comment, a nil and empty slice have the same meaning now, correct?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 28 of 28 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @RaduBerinde)

___
*[pkg/sql/opt/memo/expr.go, line 354 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcD1ua2GrtxpAj415S:-MggKb6CDvMIZKyceTEJ:byegyex) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L354)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Yeah, I should error out if any of these are combined, nice catch.
</blockquote></details>

Now that I added NoZigzagJoin, you should also add a test to make sure that it's not combined with ForceZigzag. Sorry to give you more work!
___
*[pkg/sql/opt/memo/expr_format.go, line 413 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcELIc08OUkaHbf2-D:-MggL4uM2Zrh__-jQNTE:b-58bsyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b018ef444c4c740a27e7050dffef34ed0dab759b/pkg/sql/opt/memo/expr_format.go#L413)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

This code is only reached for ScanExpr so that will never happen right?    I could print if below for ZigzagExprs, I'd just have to dupe the ScanFlags from ScanPrivate into ZigzagJoinPrivate, should I do that?
</blockquote></details>

I think it will be printed here in the case where you tried to force a zigzag but it was not successful (looks like you have a test that confirms this).

That's not a bad idea to add the flags to ZigzagJoinPrivate... although if it's only for the purpose of printing this stuff then maybe it's overkill. Up to you...
___
*[pkg/sql/opt/xform/testdata/rules/select, line 6387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MggNNj48jXK8ttISjLW:-MggNNj48jXK8ttISjLX:bxlxcu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/testdata/rules/select#L6387)):*
> ```
>            └── j @> '{"a": 1, "b": 2}'
> 
> # Generate a zigzag join on with a hint.
> ```

nit: on with a hint -> with a hint
___
*[pkg/sql/sem/tree/select.go, line 374 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MggNkfK-UsaxLjQJZjB:-MggNkfK-UsaxLjQJZjC:b-489y7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/sem/tree/select.go#L374)):*
> ```Go
> 	}
> 	if ih.ForceZigzag && ih.NoIndexJoin {
> 		return errors.New("FORCE_ZIGZAG cannot be specified in conjunction with NO_INDEX_JOIN")
> ```

also add a similar check for NoZigzagJoin


<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/zigzag_join, line 220 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHI8nQEAM2QCvSHAVM:-MgLnsyV2HMxXIUjGeIg:b86rxph) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/logictest/testdata/logic_test/zigzag_join#L220)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a test where a zig zag can't be planned like `SELECT * FROM a@{FORCE_ZIGZAG} WHERE a = 3 AND c = 'foo'`
</blockquote></details>

That's a valid query from what I can tell, did you mean if like c didn't have an index?
___
*[pkg/sql/opt/memo/expr.go, line 354 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcD1ua2GrtxpAj415S:-MhntdwDEHMw_2_XTLnb:blyphgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/memo/expr.go#L354)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Now that I added NoZigzagJoin, you should also add a test to make sure that it's not combined with ForceZigzag. Sorry to give you more work!
</blockquote></details>

No problem!
___
*[pkg/sql/opt/memo/expr_format.go, line 413 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfcELIc08OUkaHbf2-D:-MhntmB4CiwpgRpy4wFo:b-54gk9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b018ef444c4c740a27e7050dffef34ed0dab759b/pkg/sql/opt/memo/expr_format.go#L413)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think it will be printed here in the case where you tried to force a zigzag but it was not successful (looks like you have a test that confirms this).

That's not a bad idea to add the flags to ZigzagJoinPrivate... although if it's only for the purpose of printing this stuff then maybe it's overkill. Up to you...
</blockquote></details>

Seems like overkill.
___
*[pkg/sql/opt/memo/expr_format.go, line 430 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHMHwA2Ht447fMKITf:-MgM-D5R39tTjSOjgt44:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/expr_format.go#L430)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: rather than looking ahead you could print the comma before the index name for each index except the first.

```
needComma := false
for ... {
   if needComma {
       b.WriteByte(',')
    }
    b.WriteString(...)
    needComma = true
}
```
</blockquote></details>

:+1: 
___
*[pkg/sql/opt/memo/testdata/format, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHN3c25AnuXdCXSTWu:-MgLzHJn2Cbv_aihhmaY:b-loe9ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/testdata/format#L352)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

 There's no force_zigzag printed because they are only printed for scan expressions. To test your changes in `expr_format.go` you need scans to show up in the plan. You could either remove the secondary indexes from the `zigzag` table or use the `norm` directive instead of the `opt` directive so that there is no exploration performed (or both!).
</blockquote></details>

Ah, was wondering how that code would get triggered.   Should I care that the ZigzagJoinExpr doesn't print out anything related to the hint?
___
*[pkg/sql/opt/optbuilder/select.go, line 540 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHJBog-S2Pf676Jg6u:-MgM-fAdAPlAyW6QLTWr:b-fmix9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/optbuilder/select.go#L540)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Would it be simpler if the parser set `indexFlags.ForceZigzag` to true in the `FORCE_ZIGZAG=foo` case? Then the fields in these flags map directly to the fields in the AST struct.
</blockquote></details>

Good thought but doesn't really help.   I'm treating FORCE_ZIGZAG,FORCE_ZIGZAG as an error and also setting ForceZigzag to true would complicate that error handling.
___
*[pkg/sql/opt/xform/select_funcs.go, line 1093 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHKOVX3zEs6x-6ZNj0:-MhntwBZA39gp6BeaTdO:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/select_funcs.go#L1093)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can we move this into 2 checks at the top of the `ForEach` and `ForEachStartingAfter` callbacks to avoid unnecessary work?
</blockquote></details>

:+1: 
___
*[pkg/sql/opt/xform/select_funcs.go, line 1447 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHK2DtBtNX_GaTBSPA:-Mhnu09vCX0XZP7KXOLM:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/select_funcs.go#L1447)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can we put this at the top of this `ForEach` callback to avoid unnecessary work in this case?
</blockquote></details>

:+1:
___
*[pkg/sql/opt/xform/testdata/rules/join, line 8643 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MfeyNN10WbSZKTJarc_:-MgMNil11elZV1mLQfTw:bv684gz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d0e85ccf9ddaca92b87f577eeaefb4730c0f0c4/pkg/sql/opt/xform/testdata/rules/join#L8643)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like these tests still need to be moved to `pkg/sql/opt/xform/testdata/rules/select`. We have zigzag join tests there because the rules for generating zigzag joins operate on Select expressions, not on joins. 
</blockquote></details>

Boo, I moved these but forgot to push, sorry!.
___
*[pkg/sql/opt/xform/testdata/rules/select, line 6387 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MggNNj48jXK8ttISjLW:-MhnumDb-Dd4O4Qilo8w:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/testdata/rules/select#L6387)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: on with a hint -> with a hint
</blockquote></details>

:+1: 
___
*[pkg/sql/parser/sql.y, line 9577 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHF2Yq2H8W5AB7nK9v:-MgMNuG76IzeOIcmfZ7v:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/parser/sql.y#L9577)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think we should be consistent with `FORCE_INDEX` and handle a syntax like `FORCE_ZIGZAG '=' '[' iconst64 ']'` which allows forcing a specific index by its descriptor ID. There's some parser tests in `pkg/sql/parser/testdata/table_by_id` which you can add on to.
</blockquote></details>

:+1:
___
*[pkg/sql/parser/testdata/select_clauses, line 424 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHHSLXBWNLyM8jm9K5:-Mho6AhC8qS4YXgwgkEd:blajbx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/parser/testdata/select_clauses#L424)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

What is expected with `foo@{FORCE_ZIGZAG,FORCE_ZIGZAG=a}`? An error? Can you add a test for that case and maybe `foo@{FORCE_ZIGZAG=a,FORCE_ZIGZAG}` as well.
</blockquote></details>

Yeah I think it should be a "duplicate hint" kind of error.   
___
*[pkg/sql/sem/tree/select.go, line 293 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHGYkR4GF6H478qLl7:-MhnxA4M5dD3affeXpxH:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/sem/tree/select.go#L293)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Add a comment explaining what this does and how it relates to `ZigzagIndices`.
</blockquote></details>

:+1: 
___
*[pkg/sql/sem/tree/select.go, line 374 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MggNkfK-UsaxLjQJZjB:-Mho66loCVmsnGCtL6AV:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/sem/tree/select.go#L374)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

also add a similar check for NoZigzagJoin
</blockquote></details>

:+1:
___
*[pkg/sql/opt/xform/testdata/coster/join, line 998 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHKf7nC9O97gnQXQLU:-MgMNbwCDNdUPVLEXIqb:b-8fx2ut) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/xform/testdata/coster/join#L998)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: make this a little more clear like: "Verify that a huge cost is added to a scan when there is a zigzag join hint"
</blockquote></details>

:+1:


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 16 of 25 files at r6, 6 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-Mhonzgq9rKf8c7yAD3K:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
Thanks for the reviews!   We've decided not to put this in 21.2, I'll merge it in post branch for 22.1 once Marcus is back and can check it out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-Mht6UFI7go6gbos9mvt:bsbazn4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
Rebased and updated commit message, I think this is ready to go, @mgartner did I address all your concerns?   

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MjUYiFO4ggyd1xiKI6S:bxgnyzu)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/zigzag_join, line 220 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHI8nQEAM2QCvSHAVM:-MjUW_L8-GzEtvbUn4mk:b-u1koiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/logictest/testdata/logic_test/zigzag_join#L220)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

That's a valid query from what I can tell, did you mean if like c didn't have an index?
</blockquote></details>

Nevermind, I was confused.   Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 16 of 25 files at r6, 6 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-:-MjUn_Pq4sY25Jg0DV0B:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @mgartner, and @rytaft)

___
*[pkg/sql/opt/memo/testdata/format, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHN3c25AnuXdCXSTWu:-MjUlI9600ZbzUo_LKET:b-jqorho) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/testdata/format#L352)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Ah, was wondering how that code would get triggered.   Should I care that the ZigzagJoinExpr doesn't print out anything related to the hint?
</blockquote></details>

It might be a bit confusing, but I don't see it as a huge problem. I think its fine to leave as-is for now.
___
*[pkg/sql/opt/xform/select_funcs.go, line 1065 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MjUmVsNEMJwfHeDFFpu:-MjUmVsNEMJwfHeDFFpv:btbd6hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4063fe52be9f58a225591a25890af81431ab47b1/pkg/sql/opt/xform/select_funcs.go#L1065)):*
> ```Go
> 				forceIndexes := scanPrivate.Flags.ZigzagIndexes
> 				if !forceIndexes.SubsetOf(indexes) {
> 					return
> ```

nit: I think this could still go up higher to avoid unnecessary work, right below: 

```
iter2.ForEachStartingAfter(leftIndex.Ordinal(), func(rightIndex cat.Index, innerFilters memo.FiltersExpr, rightCols opt.ColSet, _ bool, _ memo.ProjectionsExpr) {
```
___
*[pkg/sql/opt/xform/select_funcs.go, line 1369 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MjUmK7P1kXwzqUy2KTv:-MjUmK7P1kXwzqUy2KTw:b-vsksb8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4063fe52be9f58a225591a25890af81431ab47b1/pkg/sql/opt/xform/select_funcs.go#L1369)):*
> ```Go
> 
> 		// Check if we have zigzag hints.
> 		if !scanPrivate.Flags.ZigzagIndexes.Empty() && !scanPrivate.Flags.ZigzagIndexes.Contains(index.Ordinal()) {
> ```

nit: I think this could still go up higher to avoid unnecessary work, right below: 

```
iter.ForEach(func(index cat.Index, filters memo.FiltersExpr, indexCols opt.ColSet, _ bool, _ memo.ProjectionsExpr) {
```
___
*[pkg/sql/opt/xform/testdata/rules/select, line 6072 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MjUmsxlC_GzMq3WzZQ_:-MjUmsxlC_GzMq3WzZQa:b-8r0te) ([raw file](https://github.com/cockroachdb/cockroach/blob/4063fe52be9f58a225591a25890af81431ab47b1/pkg/sql/opt/xform/testdata/rules/select#L6072)):*
> ```
> # Zigzag join hinting tests
> exec-ddl
> CREATE TABLE zigzag (n INT PRIMARY KEY, a INT, b INT, c STRING, INDEX a_idx(a), INDEX b_idx(b), UNIQUE INDEX c_idx(b,c), INDEX ba_idx(b));
> ```

nit: put this on multiple lines

Also, is `ba_idx` suppose to index `(b, a)`?


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> Reviewed 2 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/memo/testdata/format, line 352 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MgHN3c25AnuXdCXSTWu:-MjVNNIJB2rjp0TeCuLv:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/42c44fe01a71e47dff3430122310c2162f197c41/pkg/sql/opt/memo/testdata/format#L352)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It might be a bit confusing, but I don't see it as a huge problem. I think its fine to leave as-is for now.
</blockquote></details>

:+1:
___
*[pkg/sql/opt/xform/testdata/rules/select, line 6072 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/67737#-MjUmsxlC_GzMq3WzZQ_:-MjVNTow6N-yH2sZELiy:bkptsgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4063fe52be9f58a225591a25890af81431ab47b1/pkg/sql/opt/xform/testdata/rules/select#L6072)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: put this on multiple lines

Also, is `ba_idx` suppose to index `(b, a)`?
</blockquote></details>

I think I had it as b,a but then it didn't use zigzag, so I changed it to back to b.   But I think relying on duplicate indexes like that isn't good so I made it b,c. 


<!-- Sent from Reviewable.io -->

cucaroach commented:
Thanks for all the reviews!
bors r+
